packages/xstate-graph/src/simplePaths.ts

```diff
@@ -0,0 +1,133 @@
+  const stateMap = new Map<SerializedState, TState>();
+  const visitCtx: VisitedContext<TState, TEvent> = {
+    vertices: new Set(),
+    edges: new Set()
+  };
+  const path: any[] = [];
+  const pathMap: Record<
+    SerializedState,
+    { state: TState; paths: Array<StatePath<TState, TEvent>> }
+  > = {};
+
+  function util(
+    fromStateSerial: SerializedState,
+    toStateSerial: SerializedState
+  ) {
+    const fromState = stateMap.get(fromStateSerial)!;
+    visitCtx.vertices.add(fromStateSerial);
+
+    if (fromStateSerial === toStateSerial) {
+      if (!pathMap[toStateSerial]) {
+        pathMap[toStateSerial] = {
+          state: stateMap.get(toStateSerial)!,
+          paths: []
+        };
+      }
+
+      const toStatePlan = pathMap[toStateSerial];
+
+      const path2: StatePath<TState, TEvent> = {
+        state: fromState,
+        weight: path.length,
+        steps: [...path]
+      };
+
+      toStatePlan.paths.push(path2);
+    } else {
+      for (const serializedEvent of Object.keys(
+        adjacency[fromStateSerial].transitions
+      ) as SerializedEvent[]) {
+        const { state: nextState, event: subEvent } = adjacency[
+          fromStateSerial
+        ].transitions[serializedEvent];
+
+        if (!(serializedEvent in adjacency[fromStateSerial].transitions)) {
+          continue;
+        }
+        const prevState = stateMap.get(fromStateSerial);
+
+        const nextStateSerial = serializeState(nextState, subEvent, prevState);
+        stateMap.set(nextStateSerial, nextState);
+
+        if (!visitCtx.vertices.has(nextStateSerial)) {
+          visitCtx.edges.add(serializedEvent);
+          path.push({
+            state: stateMap.get(fromStateSerial)!,
+            event: subEvent
+          });
+          util(nextStateSerial, toStateSerial);
+        }
+      }
+    }
+
+    path.pop();
+    visitCtx.vertices.delete(fromStateSerial);
+  }
+
+  const fromStateSerial = serializeState(fromState, undefined);
+  stateMap.set(fromStateSerial, fromState);
+
+  for (const nextStateSerial of Object.keys(adjacency) as SerializedState[]) {
+    util(fromStateSerial, nextStateSerial);
+  }
+
+  const simplePaths = flatten(Object.values(pathMap).map((p) => p.paths));
+
+  if (resolvedOptions.toState) {
+    return simplePaths.filter((path) => resolvedOptions.toState!(path.state));
+  }
+
+  return simplePaths;
+}
+
+export function getSimplePathsFromTo<TState, TEvent extends EventObject>(
+  behavior: SimpleBehavior<TState, TEvent>,
+  fromPredicate: (state: TState) => boolean,
+  toPredicate: (state: TState) => boolean,
+  options: TraversalOptions<TState, TEvent>
+): Array<StatePath<TState, TEvent>> {
+  const resolvedOptions = resolveTraversalOptions(options);
+  const simplePaths = getSimplePaths(behavior, resolvedOptions);
+
+  return simplePaths.filter((path) => {
+    return (
+      toPredicate(path.state) &&
+      path.steps.some((step) => fromPredicate(step.state))
+    );
+  });
```

Comment:
I've noticed that this behaves differently from `getShortestPathsFromTo`. `getShortestPathsFromTo` first computes "from" paths and it only computes "to" paths from those. It's a better algorithm than the one here because here we have to generate **everything** and reject paths that don't match later on. 

How hard would it be to unify the approaches? Is there a reason why all simple paths have to be generated upfront for us to get an answer on the `FromTo` question?

Reply:
Improved here: 286190b51d9a9629e85e5a5244c6189a82e8ba7b